### ISSUE TITLE
iOS: onRegister only called with requestPermissions: true

### ISSUE THREAD

**Issue by drbarto** [zo0r/react-native-push-notification#1052]
Hey,

I would like to run `PushNotification.configure` on app start _without_ immediately requesting permissions from the user, therefore I set `requestPermissions` to `false`. However, in this case `onRegister` never gets called. (Neither do I get an error via `PushNotificationIOS.addEventListener`). When I set `requestPermissions` to `true`, `onRegister` is called with the token correctly.

So... 1) is this by design? and 2) is there a way to get the token without requesting permissions?

(running on an iPhone XR -- real device -- with iOS 12.2; the push entitlement is set correctly in Xcode)

**Comment by calebpanza** [comment]
I can't seem to get `onRegister` to trigger at all.

**Comment by markpar** [comment]
Same.  Triggered just fine on Android, but never called on iOS 10.x+.  Works fine on iOS 9.x.  🤔

**Comment by nolanrbrady** [comment]
I am running into this same issue. Has anyone found a cause, or better yet, a solution?

**Comment by codeoholic** [comment]
Have you enabled "Push Notifications" in Xcode - Capabilities for the application?

**Comment by CGReinhold** [comment]
I'm having the same problem after updating from **iOS** 9 to iOS 11, **react-native** 0.57.1 to 0.61.2 and **react-native-push-notification** from 3.0.2 to 3.1.9

I enabled **Push Notifications** in Xcode - Capabilities for the application and also **Remote Notifications** on the Background Modes but it still doesn't fire the onRegister event, even after the user has Allowed notifications from the default prompt.
The push notification is also enabled in the Apple developer account, since this worked in a previous version that is already published on the app store.

Here is part of my code executed on the component:

`componentDidMount() {
    PushNotification.configure({
      onRegister: token => {
        AsyncStorage.setItem('token', token.token);
      },
  
      onNotification: notification => {
        ...
      },
      senderID: 'XXXXXXXX',
      popInitialNotification: true,
      permissions: { alert: true, badge: true, sound: true },
      requestPermissions: true
    });
  }`

Am I missing something after doing this updates?

**Comment by atebit** [comment]
This is happening for me as well, just nothing for onRegister.

I've tried putting it into state, asking later,

I've tried using my own Event delegator to get to it outside of the react scope

I've tried requestPermissions:true

Push notifications cert not expired

Push Notifications capability enabled in Xcode


Last time I updated the framework this happened as well, this part seems quite buggy...

**Comment by atebit** [comment]
Also, there's no error.. Since, the code is in Pods, is there a way to log what's happening? I see the code in the Pod, but I am not going to edit it...

**Comment by atebit** [comment]
Nevermind, I had forgotten to update the AppDelegate.m:

https://github.com/react-native-community/react-native-push-notification-ios

**Comment by schumannd** [comment]
@atebit are you using RN version >0.60? If yes, why would you need to edit AppDelegate.m?

**Comment by magrinj** [comment]
@schumannd Even if you have RN version > 0.60, you still have to edit `AppDelegate.m`, auto-linking is going to edit it for you, it will just link project pods automatically. The changes reflected in `AppDelegate.m` forward notifications events to this module.

**Comment by my-name-is-nheo** [comment]
I have enabled push notifications and remote notifications. I have updated both .m and .h files as instructed (twice just incase). Changed the import format as well.  Am I missing something here?
`
`import React, {Component} from 'react';
import {StatusBar, NativeModules, Platform} from 'react-native';
import PushNotificationIOS from '@react-native-community/push-notification-ios';
import App from './AppNavigation';
import {MenuProvider} from 'react-native-popup-menu';
import NavigationService from './NavigationService';
import Amplify from 'aws-amplify';
// import Analytics from '@aws-amplify/analytics';
// import awsNotificationConfig from '../src/Constants/NotificationCredentials';
import PushNotification from '@aws-amplify/pushnotification';

const awsconfig = {
  Auth: {
 
  },
  Storage: {
   
  },
  PushNotification: {
   
  },
  API: {
    endpoints: [    ],
  },
};

Amplify.configure(awsconfig);
PushNotification.configure(awsconfig);

PushNotification.onRegister(async (token) => {
  console.log('in app registration', token);
  PushNotification.updateEndpoint(token);
});

// works on android. but not on ios.
// NativeModules.RNPushNotification.getToken((token) => {
//   console.log(`PushToken: ${token}`);
// });

PushNotification.onNotification((notification) => {
  console.log('in app notification', notification);
  if (Platform.OS === 'ios') {
    notification.finish(PushNotificationIOS.FetchResult.NoData);
  }
});

PushNotification.onNotificationOpened((notification) => {
  console.log('the notification is opened', notification);
});

if (Platform.OS === 'ios') {
  console.log(Platform.OS, 'this is platform');
  PushNotification.requestIOSPermissions();
}

export default class AppWithNavigationState extends Component<{}, {}> {
  render() {
    return (
      <MenuProvider backHandler={true}>
        <StatusBar backgroundColor="#3F51b5" barStyle="light-content" />
        <App
          ref={(navigatorRef) => {
            NavigationService.setTopLevelNavigator(navigatorRef);
          }}
        />
      </MenuProvider>
    );
  }
}
`

**Comment by dzpt** [comment]
Anyone got this? onRegister never called if requestPermissions = false.

**Comment by codeoholic** [comment]
@dzpt Did you call PushNotificationsHandler.requestPermissions() to trigger onRegister() on any other screen?

**Comment by dzpt** [comment]
@codeoholic Yes, i want to manually requestPermission, so i call it on the first time user open the app.
onRegister is called on that time

after i re-open the app, onRegister is never be called again

**Comment by codeoholic** [comment]
@dzpt onRegister gets called only once during the lifescycle. Try reinstalling the app and mange it wisely.

Here's how I manage it.
```
//App.js
PushNotification.configure({
    onRegister: function (token) {
        console.log("TOKEN:", token);
        userNotificationToken( navigationRef, token );
    },
    onRegistrationError: function(err) {
        console.log( err );
    },
    onNotification: function (notification) {
        console.log("NOTIFICATION:", notification);
        notification.finish(PushNotificationIOS.FetchResult.NoData);
    },
    permissions: {
        alert: true,
        badge: true,
        sound: true,
    },
    popInitialNotification: false,
    requestPermissions: false
});
```


```
//NotificationRequest.js
const triggerNotificationsRequest = () => {
    PushNotification.requestPermissions();
}
```
triggerNotificationsRequest will call onRegister in App.js which is higher up the order.

**Comment by dzpt** [comment]
@codeoholic i reinstall the app also, but it didn't work, i call to `requestPermissions()` once on the first launch only.

do you call `triggerNotificationsRequest` everytime open the app?

**Comment by codeoholic** [comment]
@dzpt Crisp answer, **NO**.
**Reason** : iOS guidelines only allow one time prompt for notification permission. If I try to do so, it will be blocked at OS level. So, when I prompt user for notification for the first time, I store the token in AsyncStorage. On next load, I check Async first to check if I have to show user the NotificationRequest screen or not.

**Comment by dzpt** [comment]
@codeoholic so the error is persisted. Because if you set `requestPermissions=true`, `onRegister / onRegistrationError` will be called everytime open the app. 

It makes sense for you to know the correct token to notify the backend.

**Comment by cs-manughian** [comment]
Any update on this? I cannot get PushNotificationsHandler.requestPermissions() to work for Android. No prompt shows up to ask for permissions. Same with requestPermissions: true
